mypy/plugin.py

```diff
@@ -379,3 +392,189 @@ def int_pow_callback(ctx: MethodContext) -> Type:
+        return default
+
+    def called_function(expr: Expression) -> Optional[str]:
+        if isinstance(expr, CallExpr) and isinstance(expr.callee, RefExpr):
+            return expr.callee.fullname
+        return None
+
+    decorator = ctx.reason
+
+    # Default values of attr.s()
+    init = True  # type: Optional[bool]
+    cmp = True  # type: Optional[bool]
+    auto_attribs = False  # type: Optional[bool]
+
+    if isinstance(decorator, CallExpr):
+        # Read call arguments.
+        # TODO handle these returning None (e.g. bool=SOMETHING_ELSE)
+        init = get_bool_argument(decorator, "init", init)
+        cmp = get_bool_argument(decorator, "cmp", cmp)
+        auto_attribs = get_bool_argument(decorator, "auto_attribs",
+                                         auto_attribs)
+
+    if not init and not cmp:
+        # Nothing to add.
+        return
+
+    # To support nested classes we use fullname(). But fullname is <filename>.<variable name>
+    # and named_type() expects only the name.
+    self_type = ctx.api.named_type(ctx.cls.info.fullname().split(".", 1)[1])
+    function_type = ctx.api.named_type('__builtins__.function')
+
+    if init:
+        # Walk the body looking for assignments.
+        names = []  # type: List[str]
+        types = []  # type: List[Type]
+        has_default = set()  # type: Set[str]
+
+        def add_init_argument(name: str, typ: Optional[Type], default: bool,
+                              context: Context) -> None:
+            if not default and has_default:
+                ctx.api.fail(
+                    "Non-default attributes not allowed after default attributes.",
+                    context)
+            if not typ:
+                if ctx.api.options.disallow_untyped_defs:
+                    # This is a compromise.  If you don't have a type here then the __init__ will
+                    # be untyped. But since the __init__ method doesn't have a line number it's
+                    # difficult to point to the correct line number.  So instead we just show the
+                    # error in the assignment, which is where you would fix the issue.
+                    ctx.api.fail(messages.NEED_ANNOTATION_FOR_VAR, context)
+                typ = AnyType(TypeOfAny.unannotated)
+
+            names.append(name)
+            assert typ is not None
+            types.append(typ)
+            if default:
+                has_default.add(name)
+
+        def is_class_var(expr: NameExpr) -> bool:
+            if isinstance(expr.node, Var):
+                return expr.node.is_classvar
+            return False
+
+        # Walk the mro in reverse looking for those yummy attributes.
+        for info in reversed(ctx.cls.info.mro):
+            for stmt in info.defn.defs.body:
+                if isinstance(stmt, AssignmentStmt) and isinstance(stmt.lvalues[0], NameExpr):
+                    lhs = stmt.lvalues[0]
+                    name = lhs.name.lstrip("_")
```

Comment:
Why this is necessary? Does ``attr`` has specific treatment of underscores? If yes, then please add a comment.